docs/decisions/0007-reducing-complexity.md

```diff
@@ -0,0 +1,39 @@
+# Reducing Complexity
+
+## Motivation
+
+High complexity is in general a concerning problem: It can lead to untested legacy code, a high amount of bugs, bad maintainability, slow development, and so on. These decisions are concrete changes we want to make to reduce complexity.
+
+## Decisions
+
+### Hooks
+
+Decision: Hooks always start with `use` (for example `useCounter`). This applies to custom hooks. A custom hook is any function that calls other hooks, especially state and lifecycle hooks like `useEffect` or `useState`. Functions that do not call other hooks never start with `use`.
+
+Reason: See general principles (3). See [React Docs](https://beta.reactjs.org/learn/reusing-logic-with-custom-hooks#hook-names-always-start-with-use)
+
+### Contents of hooks.js
+
+Decision: A hooks.js file lives in the directory of a React component, and must contain only custom hooks. All other functions that are not custom hooks must be in other files. (You may make an exception for functions that you consider private to that hooks.js file and do not expect to be used anywhere else.) All custom hooks that belong to the parent component must be in that hooks.js file. This does not apply for hooks that are intended to be reusable.
+
+Reason: It is common practice to have a "hooks.js" file and you will only expect hooks in there. We put all (except for reused) hooks in there in order to make it easy to find hooks and distinguish between them and normal functions.
+
+### Reusable hooks
+
+Decision: Hooks intended to be reusable live in shared directories. Either the directory name or the filename they are in must signify that they are hooks.
+
+Reason: Easily distinguish hooks from other functions
+
+### Self-imports
+
+Decision: We avoid self-imports and replace them wherever we find them.
+
+Reason: See general principles (5). They have unnecessary complexity. The conceptual idea behind it is not so clear: the way we program in JS is usually a one-directional flow where we declare a const for example, either import it or declare it in the same file, then use it down the line and export it. Conceptually the idea that you declare a constant, export it, then import it into the same file, then export again makes everything pretty convoluted, and we also have extra syntactical overhead for it. The code itself does not explain why the logic is so complicated and what it is there for - because it is only there for the tests. From a code perspective it is not logical to use such a logic. We do not need self-imports for writing tests, other projects generally do not use them.
```

Comment:
Is the broader issue here that good stewardship calls for all test-related logic to live in test files, and for production files to only have code that contributes to features? And that our use of re-importing is causing test logic to live in production files? I haven't encountered this pattern, but wonder whether it's actually a "design for testability" issue. What is it that the re-importing is enabling for tests? Is there an alternate design that keeps test logic in test files?

Reply:
Yeah there are a lot of reasons that have apparently led up to this happening. The main reason for these self-imports in our project was that if you have two functions A and B in the same file and you want to test one and mock the other out you have some problems in JS. But I have never seen this as an obstacle in other codebases, and it is so easy to just throw one of the functions into another file, which is fine from a design standpoint - considering that webpack will merge those two files together anyway, so it has no performance implications to have a bunch of files.